Optimization

jerrinsg opened the pull request:
This commit includes two optimizations:

1.Flush the network write buffer only after processing all the commands in
a pipeline. Frequent flushing was seen to be a cause of throughput
bottleneck.
2. While expanding the query buffer to accommodate more data, copy only
that data from the old buffer which has not yet been processed.

jerrinsg committed:
Make clients reuse transaction handle

Currently a new transaction handle is requested for each request that is
received by the client handler.  This is not necessary as transaction handles
can be reused and need to be released only when the client closes the
connection.

Signed-off-by: Jerrin Shaji George <jshajigeorge@vmware.com>

jerrinsg committed:
Optimizations

This commit includes two optimizations:

1.Flush the network write buffer only after processing all the commands in
a pipeline. Frequent flushing was seen to be a cause of throughput
bottleneck.
2. While expanding the query buffer to accommodate more data, copy only
that data from the old buffer which has not yet been processed.

Throughput improvements observed while running memtier benchmark:
Memtier parameters: threads=1 clients = 5 data-size=64 ratio=1:1

Pipeline    Old throughput (MB/s)   New throughput (MB/s)	Delta
1	    5.232		    5.493			4.98 %
2	    5.577		    10.299			84.66 %
8	    19.461		    25.372			30.37 %
32	    36.988		    35.885			-2.98 %
128	    38.011		    38.87			2.25 %
512	    37.918		    39.545			4.29 %
2048	    37.008		    41.684			12.63 %

Signed-off-by: Jerrin Shaji George <jshajigeorge@vmware.com>

pratapsu commented:
Jerrin: 

Always describe the test case that you used to validate this change, in your code review request, and also in the checkin message (that will make it easier to later figure out why a change made sense). In this case, you found this helped with memtier, so, say the options that you used to memtier, the throughout before your change, and after your change, and also note the hardware that you used. All this will help if we ever have to re-validate this change later.

pratapsu commented:
I don't quite understand why you raised the initial value of querybuf. Is 1K not a good starting point? Perhaps this helped some memtier incantation?

jerrinsg commented:
Throughput improvements observed with the new changes:
Memtier parameters: threads=1 clients = 5 data-size=64 ratio=1:1

| Pipeline | Old throughput | New throughput | Delta   |
|----------|----------------|----------------|---------|
|          |                |                |         |
| 1        | 5.232          | 5.493          | 4.98 %  |
| 2        | 5.577          | 10.299         | 84.66 % |
| 8        | 19.461         | 25.372         | 30.37 % |
| 32       | 36.988         | 35.885         | -2.98 % |
| 128      | 38.011         | 38.87          | 2.25 %  |
| 512      | 37.918         | 39.545         | 4.29 %  |
| 2048     | 37.008         | 41.684         | 12.63 % |

jerrinsg commented:
I actually do not see any performance improvement with increasing the size of the query buffer to 4K. So I have reverted this change and updated the commit.

pratapsu commented:
Ok, Jerrin. I have looked over these diffs, and I have no further comments.
